# Hyperparameter Tuning Pipeline

This notebook was solely run in a Google Colab T4 Python 3 environment. It took roughly 15 minutes to optimize the hyperparameters excluding the installation of dependencies.

In order to obtain the parameters used in the thesis, please run this notebook in Google Colab with a T4 Python 3 runtime.

# Install Dependencies

In [1]:
!pip install optuna==4.3.0
!pip install pandas==2.2.2
!pip install torch==2.6.0
!pip install numpy==2.0.2
!pip install matplotlib
!pip install scipy==1.15.3

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 386.6/386.6 kB 32.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 242.5/242.5 kB 19.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 111.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 99.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 61.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 105.8 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlin

# Import Libraries & Load Data

In [2]:
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
import matplotlib.pyplot as plt
from datetime import timedelta
import datetime as dt
import random
import optuna
from optuna.trial import TrialState
import os

# Read the CSV and convert date to datetime
df = pd.read_csv('./dji_etc_cc_ohlc_data.csv').drop(columns='Unnamed: 0')
df['date'] = pd.to_datetime(df['date'])
risk_free_rate = pd.read_csv('./risk_free_rate.csv')
risk_free_rate.columns = ['date', 'yield']
risk_free_rate.set_index('date', inplace=True)
risk_free_rate = risk_free_rate['yield']/100
risk_free_rate.index = pd.to_datetime(risk_free_rate.index)

# Filter to DJIA tickers drop DOW Inc. due to insufficient data
dji_ticker = ['AAPL', 'AMGN', 'AXP', 'BA', 'CAT', 'CRM', 'CSCO', 'CVX', 'DIS',
              'GS', 'HD', 'HON', 'IBM', 'INTC', 'JNJ', 'JPM', 'KO', 'MCD', 'MMM',
              'MRK', 'MSFT', 'NKE', 'PG', 'TRV', 'UNH', 'V', 'VZ', 'WBA', 'WMT',
              'IGLN.L', 'ISLN.L', 'IPLT.L', 'IPDM.L']

def set_seed(seed):
    """Set all random seeds for reproducibility."""
    random.seed(seed)          # Python
    np.random.seed(seed)       # NumPy
    torch.manual_seed(seed)    # PyTorch CPU
    torch.cuda.manual_seed(seed)       # PyTorch GPU (current CUDA device)
    torch.cuda.manual_seed_all(seed)   # PyTorch GPU (all CUDA devices)
    torch.backends.cudnn.deterministic = True  # CuDNN determinism
    torch.backends.cudnn.benchmark = False     # Disable CuDNN benchmarking

set_seed(7)

In [3]:
# Check Versions
print(f'Pandas Version: {pd.__version__}')
print(f'Numpy Version: {np.__version__}')
print(f'PyTorch Version: {torch.__version__}')
print(f'Optuna Version: {optuna.__version__}')


Pandas Version: 2.2.2
Numpy Version: 2.0.2
PyTorch Version: 2.6.0+cu124
Optuna Version: 4.3.0


# Declaration of Functions

In [4]:
# -----------------------------------------
# Data Preparation and Feature Engineering
# -----------------------------------------

def prepare_data(long_df,):
    """
    Pivot the long dataframe into a wide format with 'close' prices and calculate
    technical indicators (SMA and RSI) for each asset.
    """
    # Pivot to wide format for close prices only
    # Filter to DJIA tickers
    dji_ticker = ['AAPL', 'AMGN', 'AXP', 'BA', 'CAT', 'CRM', 'CSCO', 'CVX', 'DIS',
                'GS', 'HD', 'HON', 'IBM', 'INTC', 'JNJ', 'JPM', 'KO', 'MCD', 'MMM',
                'MRK', 'MSFT', 'NKE', 'PG', 'TRV', 'UNH', 'V', 'VZ', 'WBA', 'WMT',
                'IGLN.L', 'ISLN.L', 'IPLT.L', 'IPDM.L']

    df_filtered = long_df[long_df['tic'].isin(dji_ticker)].reset_index(drop=True)
    wide_df = df_filtered.pivot(index='date', columns='tic', values='close')
    wide_df.columns = [f'close_{col}' for col in dji_ticker]
    wide_df = wide_df.sort_index()
    wide_df = wide_df.ffill().bfill()

    # Calculate technical indicators for each ticker
    for tic in dji_ticker:
        close_col = f'close_{tic}'
        # Simple Moving Average (SMA)
        wide_df[f'sma_{tic}'] = wide_df[close_col].rolling(20).mean()

        # Relative Strength Index (RSI)
        delta = wide_df[close_col].diff()
        gain = delta.where(delta > 0, 0)
        loss = -delta.where(delta < 0, 0)
        avg_gain = gain.rolling(14).mean()
        avg_loss = loss.rolling(14).mean()
        rs = avg_gain / avg_loss
        wide_df[f'rsi_{tic}'] = 100 - (100 / (1 + rs))

    return wide_df.bfill()  # Backfill missing values

def extract_features(wide_df, rebalance_date, lookback=252):
    """
    Extract features from a lookback window ending at a given rebalance_date.
    Features include:
      - The latest SMA and RSI values (per ticker)
      - Flattened covariance matrices for the past 1-year of:
            close prices, SMAs, and RSIs.
    """
    lookback_data = wide_df.loc[:rebalance_date].iloc[-lookback:]
    # Get the list of tickers from the columns, assuming they are named like "close_<tic>"
    tics = [col.split('_')[1] for col in wide_df.columns if col.startswith('close_')]
    features = []
    # Covariance matrices for close, sma, and rsi data
    close_data = lookback_data[[f'close_{tic}' for tic in tics]].values
    sma_data = lookback_data[[f'sma_{tic}' for tic in tics]].values
    rsi_data = lookback_data[[f'rsi_{tic}' for tic in tics]].values

    for i in [close_data, sma_data, rsi_data]:
        features.append(i)

    return np.array(features)

# -------------------------
# Neural Network Definition
# -------------------------
class TimeSeriesNetwork(nn.Module):
    def __init__(self, num_tickers=33):
        super(TimeSeriesNetwork, self).__init__()
        self.conv1 = nn.Conv1d(in_channels=3, out_channels=16, kernel_size=5, padding=2)
        self.conv2 = nn.Conv1d(in_channels=16, out_channels=32, kernel_size=3, padding=1)
        self.pool = nn.AdaptiveMaxPool1d(output_size=1)

        self.fc1 = nn.Linear(in_features=32, out_features=16)
        self.fc2 = nn.Linear(in_features=16, out_features=1)  # Score per ticker

        self.num_tickers = num_tickers

    def forward(self, x):
        # x: (batch, 3, 252, num_tickers)
        batch_size = x.size(0)
        x = x.permute(0, 3, 1, 2)  # (batch, num_tickers, 3, 252)
        x = x.reshape(batch_size * self.num_tickers, 3, 252)

        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = self.pool(x).squeeze(-1)  # (batch*num_tickers, 32)

        x = F.relu(self.fc1(x))
        scores = self.fc2(x).view(batch_size, self.num_tickers)  # (batch, num_tickers)

        weights = F.softmax(scores, dim=1).unsqueeze(1)  # (batch, 1, num_tickers)
        return weights

#--------------------------
# Loss Function
#--------------------------

def sharpe_ratio_loss(returns, weights, rfr=0.024):
    """
    Compute differentiable Sharpe ratio loss.
    - returns: (batch, time, assets)
    - weights: (batch, 1, assets)
    """
    # Weighted portfolio returns per timestep
    port_returns = (returns * weights).sum(dim=2)  # (batch, time)
    daily_rfr = (1 + rfr) ** (1 / 252) - 1
    excess_returns = port_returns - daily_rfr

    mean_ret = excess_returns.mean(dim=1)
    std_ret = excess_returns.std(dim=1) + 1e-6
    sharpe = mean_ret / std_ret

    return -sharpe.mean()  # negative for minimization

def loss_function(returns, weights, rfr=0.024, reg_lambda=0.01, div_lambda=0.01):
    """
    Sharpe loss + L2 regularization + diversification penalty
    """
    sr_loss = sharpe_ratio_loss(returns, weights, rfr)

    # Regularization term
    reg = reg_lambda * (weights ** 2).sum()

    # Entropy-based diversification reward
    entropy = - (weights * torch.log(weights + 1e-6)).sum(dim=2).mean()
    div_penalty = -div_lambda * entropy

    return sr_loss + reg + div_penalty

# -------------------------
# Training Loop
# -------------------------

def train_model(model, train_features, train_targets, num_epochs=100, lr=0.001, rfr=0.024, reg_lambda=0.01, div_lambda=0.01):
    """
    Train the model on a growing dataset of past lookback windows.
    """
    model.train()
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
    dataset = TensorDataset(train_features, train_targets)
    loader = DataLoader(dataset, batch_size=16, shuffle=True)

    for epoch in range(num_epochs):
        epoch_loss = []
        for X_batch, y_batch in loader:
            X_batch = X_batch.to(device)
            y_batch = y_batch.to(device)
            weights = model(X_batch).squeeze(1)  # shape (batch, assets)
            loss = loss_function(y_batch, weights.unsqueeze(1), rfr=rfr, reg_lambda=reg_lambda, div_lambda=div_lambda)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            epoch_loss.append(loss.item())

        if (epoch + 1) % 50 == 0:
            print(f"Epoch {epoch+1}: Avg Loss = {-np.mean(epoch_loss):.4f}")

    return model


In [5]:
# ---------------------------
# Prepare Base Data and Rebalance Dates
# ---------------------------
long_df = df  # original dataframe
wide_df = prepare_data(long_df)
tics = dji_ticker

# Generate quarter ends within the date range of your DataFrame
quarter_ends = pd.date_range(start=wide_df.index.min(), end=wide_df.index.max(), freq='QE')
indices = wide_df.index.get_indexer(quarter_ends, method='pad')
valid_indices = [i for i in indices if i != -1]
adjusted_dates = wide_df.index[valid_indices]
train_end_dates = adjusted_dates[3:]
train_start_dates = adjusted_dates[:-3]

# Optimization

In [6]:

# -------------------------
# Optuna Objective Function for Hyperparameter-Tuning
# -------------------------
def objective(trial, model, train_features, train_targets, rfr):
    reg_lambda = trial.suggest_float("reg_lambda", 1e-5, 1e-1, log=True) # Suggests a number on interval [1e-1,1e-5]
    div_lambda = trial.suggest_float("div_lambda", 1e-5, 1e-1, log=True) # Suggests a number on interval [1e-1,1e-5]
    lr = trial.suggest_float("lr", 1e-5, 1e-2, log=True) # Suggests a number on interval [1e-2,1e-5]

    # Same training logic for TimeSeriesNetwork just different parameters for learning-rate, diversification and regularization lambda
    temp_model = TimeSeriesNetwork(num_tickers=model.num_tickers).to(device)
    temp_model.load_state_dict(model.state_dict())
    optimizer = torch.optim.Adam(temp_model.parameters(), lr=lr)

    dataset = TensorDataset(train_features, train_targets)
    loader = DataLoader(dataset, batch_size=16, shuffle=True)

    temp_model.train()
    # Using fewer epochs for faster runtime
    for epoch in range(20):
        total_loss = 0
        for X_batch, y_batch in loader:
            X_batch, y_batch = X_batch.to(device), y_batch.to(device)
            weights = temp_model(X_batch)
            loss = loss_function(y_batch, weights, rfr=rfr, reg_lambda=reg_lambda, div_lambda=div_lambda)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            total_loss += loss.item()

        avg_loss = total_loss / len(loader)
        trial.report(avg_loss, epoch)
        if trial.should_prune():
            raise optuna.TrialPruned()

    return avg_loss

# -------------------------
# Main Backtest Loop with Tuning
# -------------------------
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)
# Initialize data structures
optimal_params = {
    'reg_lambda': [],
    'div_lambda': [],
    'lr': []
}

model = TimeSeriesNetwork(num_tickers=len(tics)).to(device)
train_features_list = []
train_targets_list = []
allocation_weights = []
portfolio_returns = []
backtest_dates = []

# Backtest loop
for i in range(len(train_end_dates)-1):
    train_end = train_end_dates[i]
    test_start = train_end_dates[i]
    test_end = train_end_dates[i+1]

    # Prepare data
    feat = extract_features(wide_df, train_end)
    rets = wide_df[[f'close_{tic}' for tic in tics]].pct_change().fillna(0)
    rets = rets.loc[:train_end].iloc[-252:].values
    rfr = risk_free_rate.loc[:train_end].iloc[-252:].mean()

    X_tensor = torch.FloatTensor(feat).unsqueeze(0).to(device)
    y_tensor = torch.FloatTensor(rets).unsqueeze(0).to(device)

    train_features_list.append(X_tensor)
    train_targets_list.append(y_tensor)
    all_X = torch.cat(train_features_list, dim=0)
    all_y = torch.cat(train_targets_list, dim=0)

    # Hyperparameter tuning
    print(f"\n=== Tuning for {train_end.date()} ===")
    sampler = optuna.samplers.TPESampler(seed=7)
    study = optuna.create_study(direction="minimize", sampler=sampler)
    study.optimize(
        lambda trial: objective(trial, model, all_X, all_y, rfr),
        n_trials=100,
        timeout=1800,  # 30 minutes max per tuning session
        n_jobs=1,
        gc_after_trial=True
    )

    # Store optimal params
    optimal_params['reg_lambda'].append(study.best_params['reg_lambda'])
    optimal_params['div_lambda'].append(study.best_params['div_lambda'])
    optimal_params['lr'].append(study.best_params['lr'])

    # Full training with best params
    print(f"Training with params: {study.best_params}")
    model = train_model(
        model, all_X, all_y,
        num_epochs=200,
        lr=study.best_params['lr'],
        reg_lambda=study.best_params['reg_lambda'],
        div_lambda=study.best_params['div_lambda'],
        rfr=rfr
    )

    # Portfolio allocation
    with torch.no_grad():
        weights = model(X_tensor).squeeze().cpu().numpy()
        allocation_weights.append(weights)
        backtest_dates.append(test_start)

        test_returns = wide_df.loc[test_start:test_end][[f'close_{tic}' for tic in tics]].pct_change().fillna(0)
        port_ret = (test_returns.values @ weights).sum()
        portfolio_returns.append(port_ret)

# -------------------------
# Calculate Average Optimal Parameters
# -------------------------
avg_params = {
    'avg_reg_lambda': np.mean(optimal_params['reg_lambda']),
    'avg_div_lambda': np.mean(optimal_params['div_lambda']),
    'avg_lr': np.mean(optimal_params['lr'])
}

print("\n=== Average Optimal Parameters ===")
print(f"Regularization Lambda: {avg_params['avg_reg_lambda']:.6f}")
print(f"Diversification Lambda: {avg_params['avg_div_lambda']:.6f}")
print(f"Learning Rate: {avg_params['avg_lr']:.6f}")

# -------------------------
# Backtest Results
# -------------------------
backtest_results = pd.DataFrame({
    'Date': backtest_dates,
    'Weights': allocation_weights,
    'Return': portfolio_returns
})

print("\nBacktest completed!")

cuda


[I 2025-05-24 17:40:55,650] A new study created in memory with name: no-name-dd2d8e38-6fae-42d3-adf6-efdb75ae1578



=== Tuning for 2015-12-31 ===


[I 2025-05-24 17:41:01,114] Trial 0 finished with value: -0.057910602539777756 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.057910602539777756.
[I 2025-05-24 17:41:01,327] Trial 1 finished with value: -0.2932144105434418 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.2932144105434418.
[I 2025-05-24 17:41:01,542] Trial 2 finished with value: 0.021417556330561638 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.2932144105434418.
[I 2025-05-24 17:41:01,749] Trial 3 finished with value: -0.0837254449725151 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.2932144105434418.
[I 2025-05

Training with params: {'reg_lambda': 7.890841691664666e-05, 'div_lambda': 0.09999222335655504, 'lr': 0.003667213073659141}
Epoch 50: Avg Loss = 0.3699
Epoch 100: Avg Loss = 0.3708
Epoch 150: Avg Loss = 0.3707


[I 2025-05-24 17:41:19,349] A new study created in memory with name: no-name-26a2dd2a-8378-4fa8-9787-94893a15b1cd
[I 2025-05-24 17:41:19,424] Trial 0 finished with value: -0.09860704839229584 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.09860704839229584.


Epoch 200: Avg Loss = 0.3711

=== Tuning for 2016-03-31 ===


[I 2025-05-24 17:41:19,666] Trial 1 finished with value: -0.3107193112373352 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3107193112373352.
[I 2025-05-24 17:41:19,942] Trial 2 finished with value: -0.05055006593465805 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3107193112373352.
[I 2025-05-24 17:41:20,186] Trial 3 finished with value: -0.10081887245178223 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3107193112373352.
[I 2025-05-24 17:41:20,436] Trial 4 finished with value: -0.05199018120765686 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3107193112373352.
[I 2025-05-2

Training with params: {'reg_lambda': 0.002016924139699465, 'div_lambda': 0.09999218384584661, 'lr': 0.0014401584131667816}
Epoch 50: Avg Loss = 0.3721
Epoch 100: Avg Loss = 0.3725
Epoch 150: Avg Loss = 0.3726


[I 2025-05-24 17:41:43,372] A new study created in memory with name: no-name-d7004667-c804-4c10-8229-c554ad4c3512
[I 2025-05-24 17:41:43,450] Trial 0 finished with value: -0.10266005247831345 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.10266005247831345.


Epoch 200: Avg Loss = 0.3726

=== Tuning for 2016-06-30 ===


[I 2025-05-24 17:41:43,659] Trial 1 finished with value: -0.31772759556770325 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.31772759556770325.
[I 2025-05-24 17:41:43,871] Trial 2 finished with value: -0.05231831967830658 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.31772759556770325.
[I 2025-05-24 17:41:44,085] Trial 3 finished with value: -0.10269728302955627 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.31772759556770325.
[I 2025-05-24 17:41:44,287] Trial 4 finished with value: -0.05365177243947983 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.31772759556770325.
[I 2025

Training with params: {'reg_lambda': 0.0026624369156099444, 'div_lambda': 0.09965385325429374, 'lr': 0.0007492194915980664}
Epoch 50: Avg Loss = 0.3784
Epoch 100: Avg Loss = 0.3785
Epoch 150: Avg Loss = 0.3786
Epoch 200: Avg Loss = 0.3787


[I 2025-05-24 17:42:02,427] A new study created in memory with name: no-name-6d1f7e1d-b110-4c91-acd6-88d81e6942ad
[I 2025-05-24 17:42:02,511] Trial 0 finished with value: -0.11147196590900421 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.11147196590900421.



=== Tuning for 2016-09-30 ===


[I 2025-05-24 17:42:02,735] Trial 1 finished with value: -0.32667630910873413 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.32667630910873413.
[I 2025-05-24 17:42:02,937] Trial 2 finished with value: -0.06109926104545593 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.32667630910873413.
[I 2025-05-24 17:42:03,143] Trial 3 finished with value: -0.10819053649902344 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.32667630910873413.
[I 2025-05-24 17:42:03,345] Trial 4 finished with value: -0.06245458498597145 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.32667630910873413.
[I 2025

Training with params: {'reg_lambda': 0.0017960147321589065, 'div_lambda': 0.09999222335655504, 'lr': 1.7309034719370482e-05}
Epoch 50: Avg Loss = 0.3884
Epoch 100: Avg Loss = 0.3885


[I 2025-05-24 17:42:21,326] A new study created in memory with name: no-name-e6bb745d-99ad-4654-8a8b-bb8b427346d7


Epoch 150: Avg Loss = 0.3887
Epoch 200: Avg Loss = 0.3888

=== Tuning for 2016-12-30 ===


[I 2025-05-24 17:42:21,427] Trial 0 finished with value: -0.1251668632030487 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.1251668632030487.
[I 2025-05-24 17:42:21,631] Trial 1 finished with value: -0.34167957305908203 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.34167957305908203.
[I 2025-05-24 17:42:21,827] Trial 2 finished with value: -0.07403574883937836 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.34167957305908203.
[I 2025-05-24 17:42:22,018] Trial 3 finished with value: -0.12097164243459702 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.34167957305908203.
[I 2025-0

Training with params: {'reg_lambda': 0.0008020247295098106, 'div_lambda': 0.09999222335655504, 'lr': 0.0007778434283793242}
Epoch 50: Avg Loss = 0.4050
Epoch 100: Avg Loss = 0.4055
Epoch 150: Avg Loss = 0.4058


[I 2025-05-24 17:42:41,455] A new study created in memory with name: no-name-ab958c74-d6ec-4538-b6eb-8615c396051b
[I 2025-05-24 17:42:41,562] Trial 0 finished with value: -0.13883014023303986 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.13883014023303986.


Epoch 200: Avg Loss = 0.4060

=== Tuning for 2017-03-31 ===


[I 2025-05-24 17:42:41,780] Trial 1 finished with value: -0.35505667328834534 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.35505667328834534.
[I 2025-05-24 17:42:41,991] Trial 2 finished with value: -0.08882158249616623 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.35505667328834534.
[I 2025-05-24 17:42:42,216] Trial 3 finished with value: -0.13001635670661926 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.35505667328834534.
[I 2025-05-24 17:42:42,445] Trial 4 finished with value: -0.08963443338871002 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.35505667328834534.
[I 2025

Training with params: {'reg_lambda': 0.00042640470885025884, 'div_lambda': 0.09973689195461088, 'lr': 0.0013702947411811492}
Epoch 50: Avg Loss = 0.4172
Epoch 100: Avg Loss = 0.4176


[I 2025-05-24 17:42:59,779] A new study created in memory with name: no-name-14dac62c-cef3-41a3-9264-38f1798d59b0


Epoch 150: Avg Loss = 0.4178
Epoch 200: Avg Loss = 0.4180

=== Tuning for 2017-06-30 ===


[I 2025-05-24 17:42:59,856] Trial 0 finished with value: -0.15601539611816406 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.15601539611816406.
[I 2025-05-24 17:43:00,072] Trial 1 finished with value: -0.370074987411499 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.370074987411499.
[I 2025-05-24 17:43:00,274] Trial 2 finished with value: -0.10836619138717651 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.370074987411499.
[I 2025-05-24 17:43:00,469] Trial 3 finished with value: -0.14186815917491913 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.370074987411499.
[I 2025-05-24 1

Training with params: {'reg_lambda': 4.079141644079937e-05, 'div_lambda': 0.09999222335655504, 'lr': 0.0005335579997338555}
Epoch 50: Avg Loss = 0.4337
Epoch 100: Avg Loss = 0.4341
Epoch 150: Avg Loss = 0.4343


[I 2025-05-24 17:43:19,129] A new study created in memory with name: no-name-0ee36e92-d37a-4688-a26a-bd2991204b8d
[I 2025-05-24 17:43:19,225] Trial 0 finished with value: -0.17295072972774506 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.17295072972774506.


Epoch 200: Avg Loss = 0.4345

=== Tuning for 2017-09-29 ===


[I 2025-05-24 17:43:19,493] Trial 1 finished with value: -0.3886040449142456 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3886040449142456.
[I 2025-05-24 17:43:19,759] Trial 2 finished with value: -0.12730219960212708 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3886040449142456.
[I 2025-05-24 17:43:20,032] Trial 3 finished with value: -0.15798905491828918 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3886040449142456.
[I 2025-05-24 17:43:20,323] Trial 4 finished with value: -0.12810659408569336 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3886040449142456.
[I 2025-05-2

Training with params: {'reg_lambda': 5.812479497892581e-05, 'div_lambda': 0.09994031805277913, 'lr': 0.00038563414461794933}
Epoch 50: Avg Loss = 0.4522
Epoch 100: Avg Loss = 0.4526


[I 2025-05-24 17:43:38,590] A new study created in memory with name: no-name-0c257d39-9869-4fff-9d74-f5d90deab75a


Epoch 150: Avg Loss = 0.4529
Epoch 200: Avg Loss = 0.4531

=== Tuning for 2017-12-29 ===


[I 2025-05-24 17:43:38,672] Trial 0 finished with value: -0.19421467185020447 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.19421467185020447.
[I 2025-05-24 17:43:38,889] Trial 1 finished with value: -0.4098894000053406 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.4098894000053406.
[I 2025-05-24 17:43:39,106] Trial 2 finished with value: -0.14894205331802368 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.4098894000053406.
[I 2025-05-24 17:43:39,315] Trial 3 finished with value: -0.16206152737140656 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.4098894000053406.
[I 2025-05-

Training with params: {'reg_lambda': 5.812479497892581e-05, 'div_lambda': 0.09994031805277913, 'lr': 0.00038563414461794933}
Epoch 50: Avg Loss = 0.4742
Epoch 100: Avg Loss = 0.4747
Epoch 150: Avg Loss = 0.4751


[I 2025-05-24 17:43:58,099] A new study created in memory with name: no-name-db22b8b6-664a-4207-8309-3623f93610aa
[I 2025-05-24 17:43:58,203] Trial 0 finished with value: -0.19704535603523254 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.19704535603523254.


Epoch 200: Avg Loss = 0.4754

=== Tuning for 2018-03-29 ===


[I 2025-05-24 17:43:58,472] Trial 1 finished with value: -0.40984204411506653 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.40984204411506653.
[I 2025-05-24 17:43:58,716] Trial 2 finished with value: -0.15166933834552765 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.40984204411506653.
[I 2025-05-24 17:43:58,967] Trial 3 finished with value: -0.16163994371891022 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.40984204411506653.
[I 2025-05-24 17:43:59,259] Trial 4 finished with value: -0.15240636467933655 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.40984204411506653.
[I 2025

Training with params: {'reg_lambda': 0.00017225050156613862, 'div_lambda': 0.09945080852136436, 'lr': 0.00040420120950699625}
Epoch 50: Avg Loss = 0.4721
Epoch 100: Avg Loss = 0.4724
Epoch 150: Avg Loss = 0.4726


[I 2025-05-24 17:44:17,506] A new study created in memory with name: no-name-86983191-ff88-4f85-8ccd-ac9765266636
[I 2025-05-24 17:44:17,593] Trial 0 finished with value: -0.19500228762626648 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.19500228762626648.


Epoch 200: Avg Loss = 0.4727

=== Tuning for 2018-06-29 ===


[I 2025-05-24 17:44:17,813] Trial 1 finished with value: -0.4068150520324707 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.4068150520324707.
[I 2025-05-24 17:44:18,038] Trial 2 finished with value: -0.15090341866016388 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.4068150520324707.
[I 2025-05-24 17:44:18,243] Trial 3 finished with value: -0.15194788575172424 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.4068150520324707.
[I 2025-05-24 17:44:18,461] Trial 4 finished with value: -0.15179048478603363 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.4068150520324707.
[I 2025-05-2

Training with params: {'reg_lambda': 5.812479497892581e-05, 'div_lambda': 0.09994031805277913, 'lr': 0.00038563414461794933}
Epoch 50: Avg Loss = 0.4713
Epoch 100: Avg Loss = 0.4715
Epoch 150: Avg Loss = 0.4717


[I 2025-05-24 17:44:36,394] A new study created in memory with name: no-name-06f123f1-e126-4a32-81a2-7350aa2bf6cd
[I 2025-05-24 17:44:36,495] Trial 0 finished with value: -0.19351667165756226 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.19351667165756226.


Epoch 200: Avg Loss = 0.4718

=== Tuning for 2018-09-28 ===


[I 2025-05-24 17:44:36,790] Trial 1 finished with value: -0.4043999910354614 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.4043999910354614.
[I 2025-05-24 17:44:37,076] Trial 2 finished with value: -0.14931756258010864 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.4043999910354614.
[I 2025-05-24 17:44:37,342] Trial 3 finished with value: -0.15614348649978638 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.4043999910354614.
[I 2025-05-24 17:44:37,592] Trial 4 finished with value: -0.15015947818756104 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.4043999910354614.
[I 2025-05-2

Training with params: {'reg_lambda': 1.2596755499034368e-05, 'div_lambda': 0.09999218027675966, 'lr': 0.0007136118280016492}
Epoch 50: Avg Loss = 0.4696
Epoch 100: Avg Loss = 0.4699
Epoch 150: Avg Loss = 0.4700


[I 2025-05-24 17:44:56,025] A new study created in memory with name: no-name-04bcd073-c44c-479b-bca8-043fcde13fa5
[I 2025-05-24 17:44:56,111] Trial 0 finished with value: -0.18464551866054535 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18464551866054535.


Epoch 200: Avg Loss = 0.4701

=== Tuning for 2018-12-31 ===


[I 2025-05-24 17:44:56,329] Trial 1 finished with value: -0.3950660228729248 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3950660228729248.
[I 2025-05-24 17:44:56,555] Trial 2 finished with value: -0.1400236189365387 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3950660228729248.
[I 2025-05-24 17:44:56,765] Trial 3 finished with value: -0.1532972753047943 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3950660228729248.
[I 2025-05-24 17:44:56,966] Trial 4 finished with value: -0.14095379412174225 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3950660228729248.
[I 2025-05-24 

Training with params: {'reg_lambda': 2.9410641048210416e-05, 'div_lambda': 0.09999223542239047, 'lr': 9.676032429406819e-05}
Epoch 50: Avg Loss = 0.4607
Epoch 100: Avg Loss = 0.4608
Epoch 150: Avg Loss = 0.4608


[I 2025-05-24 17:45:15,358] A new study created in memory with name: no-name-fb0ec445-4a24-4d8d-b539-16577b2885e6
[I 2025-05-24 17:45:15,459] Trial 0 finished with value: -0.18390077352523804 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18390077352523804.


Epoch 200: Avg Loss = 0.4608

=== Tuning for 2019-03-29 ===


[I 2025-05-24 17:45:15,750] Trial 1 finished with value: -0.3939136564731598 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3939136564731598.
[I 2025-05-24 17:45:16,028] Trial 2 finished with value: -0.13891133666038513 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3939136564731598.
[I 2025-05-24 17:45:16,292] Trial 3 finished with value: -0.1528072953224182 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3939136564731598.
[I 2025-05-24 17:45:16,548] Trial 4 finished with value: -0.14012256264686584 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3939136564731598.
[I 2025-05-24

Training with params: {'reg_lambda': 0.00010711410171454724, 'div_lambda': 0.09994031805277913, 'lr': 0.00013187065370718015}
Epoch 50: Avg Loss = 0.4589
Epoch 100: Avg Loss = 0.4589
Epoch 150: Avg Loss = 0.4590


[I 2025-05-24 17:45:35,274] A new study created in memory with name: no-name-69a279f9-c21c-4598-9b81-61eb5a013385
[I 2025-05-24 17:45:35,365] Trial 0 finished with value: -0.18508493900299072 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18508493900299072.


Epoch 200: Avg Loss = 0.4591

=== Tuning for 2019-06-28 ===


[I 2025-05-24 17:45:35,596] Trial 1 finished with value: -0.39321771264076233 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.39321771264076233.
[I 2025-05-24 17:45:35,814] Trial 2 finished with value: -0.13915181159973145 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.39321771264076233.
[I 2025-05-24 17:45:36,048] Trial 3 finished with value: -0.1511165052652359 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.39321771264076233.
[I 2025-05-24 17:45:36,271] Trial 4 finished with value: -0.1405213177204132 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.39321771264076233.
[I 2025-0

Training with params: {'reg_lambda': 2.9410641048210416e-05, 'div_lambda': 0.09999223542239047, 'lr': 9.676032429406819e-05}
Epoch 50: Avg Loss = 0.4587
Epoch 100: Avg Loss = 0.4589
Epoch 150: Avg Loss = 0.4590


[I 2025-05-24 17:45:54,661] A new study created in memory with name: no-name-d5d68818-d3e6-4de5-8b95-7d6b22f2b7be
[I 2025-05-24 17:45:54,803] Trial 0 finished with value: -0.18604736030101776 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18604736030101776.


Epoch 200: Avg Loss = 0.4591

=== Tuning for 2019-09-30 ===


[I 2025-05-24 17:45:55,094] Trial 1 finished with value: -0.39151573181152344 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.39151573181152344.
[I 2025-05-24 17:45:55,368] Trial 2 finished with value: -0.14056715369224548 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.39151573181152344.
[I 2025-05-24 17:45:55,653] Trial 3 finished with value: -0.16085635125637054 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.39151573181152344.
[I 2025-05-24 17:45:55,941] Trial 4 finished with value: -0.14198996126651764 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.39151573181152344.
[I 2025

Training with params: {'reg_lambda': 0.0001601854542501331, 'div_lambda': 0.09994031805277913, 'lr': 0.00013187065370718015}
Epoch 50: Avg Loss = 0.4580
Epoch 100: Avg Loss = 0.4581
Epoch 150: Avg Loss = 0.4583


[I 2025-05-24 17:46:14,458] A new study created in memory with name: no-name-4ebf2262-2373-45a4-8e96-a5cbdf6f5bfd
[I 2025-05-24 17:46:14,611] Trial 0 finished with value: -0.18194038420915604 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18194038420915604.


Epoch 200: Avg Loss = 0.4584

=== Tuning for 2019-12-31 ===


[I 2025-05-24 17:46:14,903] Trial 1 finished with value: -0.38415899872779846 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.38415899872779846.
[I 2025-05-24 17:46:15,183] Trial 2 finished with value: -0.1522769331932068 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.38415899872779846.
[I 2025-05-24 17:46:15,472] Trial 3 finished with value: -0.12681934982538223 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.38415899872779846.
[I 2025-05-24 17:46:15,761] Trial 4 finished with value: -0.10951811447739601 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.38415899872779846.
[I 2025-

Training with params: {'reg_lambda': 0.0035198299032672352, 'div_lambda': 0.09881923082471136, 'lr': 0.0008286218458794521}
Epoch 50: Avg Loss = 0.4874
Epoch 100: Avg Loss = 0.4672
Epoch 150: Avg Loss = 0.4663


[I 2025-05-24 17:46:37,401] A new study created in memory with name: no-name-2a9f59a2-1793-4192-94eb-0db65264fbed
[I 2025-05-24 17:46:37,552] Trial 0 finished with value: -0.20373767614364624 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.20373767614364624.


Epoch 200: Avg Loss = 0.4472

=== Tuning for 2020-03-31 ===


[I 2025-05-24 17:46:37,842] Trial 1 finished with value: -0.37647363543510437 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.37647363543510437.
[I 2025-05-24 17:46:38,128] Trial 2 finished with value: -0.14709152281284332 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.37647363543510437.
[I 2025-05-24 17:46:38,419] Trial 3 finished with value: -0.12797975540161133 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.37647363543510437.
[I 2025-05-24 17:46:38,700] Trial 4 finished with value: -0.18272411823272705 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.37647363543510437.
[I 2025

Training with params: {'reg_lambda': 0.00015892473241640786, 'div_lambda': 0.0985816276063012, 'lr': 0.001413307497863623}
Epoch 50: Avg Loss = 0.4971
Epoch 100: Avg Loss = 0.4567
Epoch 150: Avg Loss = 0.4087


[I 2025-05-24 17:46:59,464] A new study created in memory with name: no-name-7e6f73e0-46e0-47b5-a3a0-b31f111161cb


Epoch 200: Avg Loss = 0.4390

=== Tuning for 2020-06-30 ===


[I 2025-05-24 17:46:59,676] Trial 0 finished with value: -0.1966775730252266 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.1966775730252266.
[I 2025-05-24 17:47:00,028] Trial 1 finished with value: -0.37856072187423706 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.37856072187423706.
[I 2025-05-24 17:47:00,386] Trial 2 finished with value: -0.1910771131515503 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.37856072187423706.
[I 2025-05-24 17:47:00,722] Trial 3 finished with value: -0.148432657122612 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.37856072187423706.
[I 2025-05-2

Training with params: {'reg_lambda': 0.0008129017862960982, 'div_lambda': 0.09910473685034855, 'lr': 0.0018661206928818415}
Epoch 50: Avg Loss = 0.4451
Epoch 100: Avg Loss = 0.4463
Epoch 150: Avg Loss = 0.4820


[I 2025-05-24 17:47:21,495] A new study created in memory with name: no-name-57f17362-dbb8-4091-b549-eac35a2411c2
[I 2025-05-24 17:47:21,649] Trial 0 finished with value: -0.19203577935695648 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.19203577935695648.


Epoch 200: Avg Loss = 0.4415

=== Tuning for 2020-09-30 ===


[I 2025-05-24 17:47:21,943] Trial 1 finished with value: -0.3764231950044632 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3764231950044632.
[I 2025-05-24 17:47:22,240] Trial 2 finished with value: -0.1668330542743206 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3764231950044632.
[I 2025-05-24 17:47:22,527] Trial 3 finished with value: -0.14544130116701126 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3764231950044632.
[I 2025-05-24 17:47:22,817] Trial 4 finished with value: -0.1339813694357872 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3764231950044632.
[I 2025-05-24 

Training with params: {'reg_lambda': 4.110789070484602e-05, 'div_lambda': 0.09906255492796999, 'lr': 0.000423380174073879}
Epoch 50: Avg Loss = 0.4506
Epoch 100: Avg Loss = 0.4400
Epoch 150: Avg Loss = 0.4576


[I 2025-05-24 17:47:44,504] A new study created in memory with name: no-name-dc175f6e-b1e5-45b6-96e4-79a6c9c7fdc2
[I 2025-05-24 17:47:44,657] Trial 0 finished with value: -0.17206892371177673 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.17206892371177673.


Epoch 200: Avg Loss = 0.4404

=== Tuning for 2020-12-31 ===


[I 2025-05-24 17:47:44,942] Trial 1 finished with value: -0.3942694365978241 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3942694365978241.
[I 2025-05-24 17:47:45,230] Trial 2 finished with value: -0.1462920978665352 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3942694365978241.
[I 2025-05-24 17:47:45,523] Trial 3 finished with value: -0.15107131004333496 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3942694365978241.
[I 2025-05-24 17:47:45,803] Trial 4 finished with value: -0.15729117393493652 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3942694365978241.
[I 2025-05-24

Training with params: {'reg_lambda': 0.00016124080190496539, 'div_lambda': 0.09904864755707724, 'lr': 0.0005909336345206725}
Epoch 50: Avg Loss = 0.4436
Epoch 100: Avg Loss = 0.4368
Epoch 150: Avg Loss = 0.4573


[I 2025-05-24 17:48:06,957] A new study created in memory with name: no-name-04fc3792-b3be-4b98-8e44-7f98746b9a5b
[I 2025-05-24 17:48:07,115] Trial 0 finished with value: -0.19860897958278656 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.19860897958278656.


Epoch 200: Avg Loss = 0.4644

=== Tuning for 2021-03-31 ===


[I 2025-05-24 17:48:07,408] Trial 1 finished with value: -0.4102979153394699 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.4102979153394699.
[I 2025-05-24 17:48:07,700] Trial 2 finished with value: -0.1447123885154724 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.4102979153394699.
[I 2025-05-24 17:48:07,996] Trial 3 finished with value: -0.16401579976081848 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.4102979153394699.
[I 2025-05-24 17:48:08,286] Trial 4 finished with value: -0.1298464834690094 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.4102979153394699.
[I 2025-05-24 

Training with params: {'reg_lambda': 0.01050951146138329, 'div_lambda': 0.09973694425895253, 'lr': 0.00020839657024705375}
Epoch 50: Avg Loss = 0.4630
Epoch 100: Avg Loss = 0.4565
Epoch 150: Avg Loss = 0.4453


[I 2025-05-24 17:48:29,034] A new study created in memory with name: no-name-637aa979-f97a-4ef4-ad2b-1c81e735797d
[I 2025-05-24 17:48:29,190] Trial 0 finished with value: -0.18817681819200516 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18817681819200516.


Epoch 200: Avg Loss = 0.4434

=== Tuning for 2021-06-30 ===


[I 2025-05-24 17:48:29,484] Trial 1 finished with value: -0.4008634388446808 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.4008634388446808.
[I 2025-05-24 17:48:29,829] Trial 2 finished with value: -0.14529545605182648 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.4008634388446808.
[I 2025-05-24 17:48:30,183] Trial 3 finished with value: -0.15964708477258682 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.4008634388446808.
[I 2025-05-24 17:48:30,538] Trial 4 finished with value: -0.13299798592925072 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.4008634388446808.
[I 2025-05-2

Training with params: {'reg_lambda': 0.014820229648760062, 'div_lambda': 0.099364353125051, 'lr': 0.0009668491506587649}
Epoch 50: Avg Loss = 0.4425
Epoch 100: Avg Loss = 0.4519
Epoch 150: Avg Loss = 0.4608


[I 2025-05-24 17:48:51,903] A new study created in memory with name: no-name-f7eefbe9-ff48-40a5-90eb-e360d427cdca
[I 2025-05-24 17:48:52,064] Trial 0 finished with value: -0.18247362971305847 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18247362971305847.


Epoch 200: Avg Loss = 0.4435

=== Tuning for 2021-09-30 ===


[I 2025-05-24 17:48:52,360] Trial 1 finished with value: -0.3936179578304291 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3936179578304291.
[I 2025-05-24 17:48:52,654] Trial 2 finished with value: -0.14340399205684662 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3936179578304291.
[I 2025-05-24 17:48:52,965] Trial 3 finished with value: -0.15743286162614822 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3936179578304291.
[I 2025-05-24 17:48:53,260] Trial 4 finished with value: -0.14110591262578964 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3936179578304291.
[I 2025-05-2

Training with params: {'reg_lambda': 1.3029420032563234e-05, 'div_lambda': 0.09904864827101478, 'lr': 0.0005909336345206725}
Epoch 50: Avg Loss = 0.4478
Epoch 100: Avg Loss = 0.4421
Epoch 150: Avg Loss = 0.4518


[I 2025-05-24 17:49:14,703] A new study created in memory with name: no-name-f5062ced-c91f-4c43-88b7-d9fbaa35771b
[I 2025-05-24 17:49:14,866] Trial 0 finished with value: -0.18755222856998444 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.18755222856998444.


Epoch 200: Avg Loss = 0.4519

=== Tuning for 2021-12-31 ===


[I 2025-05-24 17:49:15,164] Trial 1 finished with value: -0.39423342049121857 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.39423342049121857.
[I 2025-05-24 17:49:15,460] Trial 2 finished with value: -0.14123564213514328 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.39423342049121857.
[I 2025-05-24 17:49:15,768] Trial 3 finished with value: -0.15461790561676025 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.39423342049121857.
[I 2025-05-24 17:49:16,064] Trial 4 finished with value: -0.14590231329202652 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.39423342049121857.
[I 2025

Training with params: {'reg_lambda': 0.00010502955079025406, 'div_lambda': 0.09958830378790556, 'lr': 0.00010826491194637938}
Epoch 50: Avg Loss = 0.4624
Epoch 100: Avg Loss = 0.4489
Epoch 150: Avg Loss = 0.4525


[I 2025-05-24 17:49:37,201] A new study created in memory with name: no-name-b8961a16-b645-43af-b46e-ae43e8c4e027


Epoch 200: Avg Loss = 0.4591

=== Tuning for 2022-03-31 ===


[I 2025-05-24 17:49:37,422] Trial 0 finished with value: -0.1875089630484581 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.1875089630484581.
[I 2025-05-24 17:49:37,801] Trial 1 finished with value: -0.38495439291000366 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.38495439291000366.
[I 2025-05-24 17:49:38,099] Trial 2 finished with value: -0.14274626970291138 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.38495439291000366.
[I 2025-05-24 17:49:38,395] Trial 3 finished with value: -0.15346073359251022 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.38495439291000366.
[I 2025-0

Training with params: {'reg_lambda': 2.14127899373953e-05, 'div_lambda': 0.09945082475171017, 'lr': 0.0008478576208587823}
Epoch 50: Avg Loss = 0.4531
Epoch 100: Avg Loss = 0.4558
Epoch 150: Avg Loss = 0.4554


[I 2025-05-24 17:49:59,347] A new study created in memory with name: no-name-100967b9-1600-4516-8b63-c8bcfbc041b7
[I 2025-05-24 17:49:59,512] Trial 0 finished with value: -0.1766260340809822 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.1766260340809822.


Epoch 200: Avg Loss = 0.4502

=== Tuning for 2022-06-30 ===


[I 2025-05-24 17:49:59,812] Trial 1 finished with value: -0.3835785835981369 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3835785835981369.
[I 2025-05-24 17:50:00,103] Trial 2 finished with value: -0.13685721531510353 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3835785835981369.
[I 2025-05-24 17:50:00,408] Trial 3 finished with value: -0.14708838611841202 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3835785835981369.
[I 2025-05-24 17:50:00,706] Trial 4 finished with value: -0.1400342509150505 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3835785835981369.
[I 2025-05-24

Training with params: {'reg_lambda': 0.00013845391171911833, 'div_lambda': 0.09999222335655504, 'lr': 0.0005288826143364468}
Epoch 50: Avg Loss = 0.4512
Epoch 100: Avg Loss = 0.4435
Epoch 150: Avg Loss = 0.4505


[I 2025-05-24 17:50:22,593] A new study created in memory with name: no-name-8195b122-186d-4e6f-b4b8-208388a7c330
[I 2025-05-24 17:50:22,759] Trial 0 finished with value: -0.17523233592510223 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.17523233592510223.


Epoch 200: Avg Loss = 0.4466

=== Tuning for 2022-09-30 ===


[I 2025-05-24 17:50:23,066] Trial 1 finished with value: -0.3781801462173462 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3781801462173462.
[I 2025-05-24 17:50:23,378] Trial 2 finished with value: -0.13216286152601242 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3781801462173462.
[I 2025-05-24 17:50:23,678] Trial 3 finished with value: -0.14404785633087158 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3781801462173462.
[I 2025-05-24 17:50:23,975] Trial 4 finished with value: -0.13297294825315475 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3781801462173462.
[I 2025-05-2

Training with params: {'reg_lambda': 0.00014587892881114746, 'div_lambda': 0.09906253691693358, 'lr': 7.615378507354793e-05}
Epoch 50: Avg Loss = 0.4379
Epoch 100: Avg Loss = 0.4423
Epoch 150: Avg Loss = 0.4365


[I 2025-05-24 17:50:45,768] A new study created in memory with name: no-name-4875b34b-09f0-4457-b194-ed1cca4d0ed1
[I 2025-05-24 17:50:45,943] Trial 0 finished with value: -0.17293164879083633 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.17293164879083633.


Epoch 200: Avg Loss = 0.4354

=== Tuning for 2022-12-30 ===


[I 2025-05-24 17:50:46,252] Trial 1 finished with value: -0.36993570625782013 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.36993570625782013.
[I 2025-05-24 17:50:46,554] Trial 2 finished with value: -0.1264117807149887 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.36993570625782013.
[I 2025-05-24 17:50:46,857] Trial 3 finished with value: -0.13534469157457352 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.36993570625782013.
[I 2025-05-24 17:50:47,159] Trial 4 finished with value: -0.12961260601878166 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.36993570625782013.
[I 2025-

Training with params: {'reg_lambda': 0.00016124080190496539, 'div_lambda': 0.09904864755707724, 'lr': 0.0005909336345206725}
Epoch 50: Avg Loss = 0.4311
Epoch 100: Avg Loss = 0.4346
Epoch 150: Avg Loss = 0.4311


[I 2025-05-24 17:51:08,570] A new study created in memory with name: no-name-a30d27ac-5ef9-4655-8354-04dd6626fae9


Epoch 200: Avg Loss = 0.4318

=== Tuning for 2023-03-31 ===


[I 2025-05-24 17:51:08,800] Trial 0 finished with value: -0.16642921417951584 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.16642921417951584.
[I 2025-05-24 17:51:09,155] Trial 1 finished with value: -0.3635449707508087 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3635449707508087.
[I 2025-05-24 17:51:09,465] Trial 2 finished with value: -0.12313875183463097 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3635449707508087.
[I 2025-05-24 17:51:09,780] Trial 3 finished with value: -0.13778158277273178 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3635449707508087.
[I 2025-05-

Training with params: {'reg_lambda': 0.0003868012377291768, 'div_lambda': 0.09973844660819883, 'lr': 0.0007976294558561055}
Epoch 50: Avg Loss = 0.4294
Epoch 100: Avg Loss = 0.4309
Epoch 150: Avg Loss = 0.4295


[I 2025-05-24 17:51:30,649] A new study created in memory with name: no-name-312726ec-90e8-41cb-9b87-37530456bfc1
[I 2025-05-24 17:51:30,823] Trial 0 finished with value: -0.1631920486688614 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.1631920486688614.


Epoch 200: Avg Loss = 0.4284

=== Tuning for 2023-06-30 ===


[I 2025-05-24 17:51:31,141] Trial 1 finished with value: -0.3627820461988449 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3627820461988449.
[I 2025-05-24 17:51:31,448] Trial 2 finished with value: -0.12152747064828873 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3627820461988449.
[I 2025-05-24 17:51:31,759] Trial 3 finished with value: -0.1382337287068367 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3627820461988449.
[I 2025-05-24 17:51:32,113] Trial 4 finished with value: -0.12305370345711708 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.3627820461988449.
[I 2025-05-24

Training with params: {'reg_lambda': 0.00016124080190496539, 'div_lambda': 0.09904864755707724, 'lr': 0.0005909336345206725}
Epoch 50: Avg Loss = 0.4259
Epoch 100: Avg Loss = 0.4257
Epoch 150: Avg Loss = 0.4256


[I 2025-05-24 17:51:53,402] A new study created in memory with name: no-name-8f5b0fac-1ffa-48a7-a9a7-9f70b462ccf0
[I 2025-05-24 17:51:53,579] Trial 0 finished with value: -0.16145353019237518 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.16145353019237518.


Epoch 200: Avg Loss = 0.4258

=== Tuning for 2023-09-29 ===


[I 2025-05-24 17:51:53,901] Trial 1 finished with value: -0.36109842360019684 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.36109842360019684.
[I 2025-05-24 17:51:54,212] Trial 2 finished with value: -0.11875863373279572 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.36109842360019684.
[I 2025-05-24 17:51:54,519] Trial 3 finished with value: -0.13915769010782242 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.36109842360019684.
[I 2025-05-24 17:51:54,825] Trial 4 finished with value: -0.12136918678879738 and parameters: {'reg_lambda': 0.000334013894021542, 'div_lambda': 1.8354619576988663e-05, 'lr': 7.318748070568251e-05}. Best is trial 1 with value: -0.36109842360019684.
[I 2025

Training with params: {'reg_lambda': 0.00010711410171454724, 'div_lambda': 0.09994031805277913, 'lr': 0.00013187065370718015}
Epoch 50: Avg Loss = 0.4268
Epoch 100: Avg Loss = 0.4269
Epoch 150: Avg Loss = 0.4270


[I 2025-05-24 17:52:17,222] A new study created in memory with name: no-name-40e198fb-c2bd-40ee-8774-becf0c6c93f7


Epoch 200: Avg Loss = 0.4270

=== Tuning for 2023-12-29 ===


[I 2025-05-24 17:52:17,465] Trial 0 finished with value: -0.14629584550857544 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.14629584550857544.
[I 2025-05-24 17:52:17,842] Trial 1 finished with value: -0.38633865118026733 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.38633865118026733.
[I 2025-05-24 17:52:18,211] Trial 2 finished with value: -0.10592246055603027 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.38633865118026733.
[I 2025-05-24 17:52:18,585] Trial 3 finished with value: -0.1307955409089724 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.38633865118026733.
[I 2025-

Training with params: {'reg_lambda': 0.010052540238067825, 'div_lambda': 0.09945176051700999, 'lr': 0.0029339647104769874}
Epoch 50: Avg Loss = 0.4353
Epoch 100: Avg Loss = 0.4791
Epoch 150: Avg Loss = 0.4357


[I 2025-05-24 17:52:41,498] A new study created in memory with name: no-name-c5465a0d-e96a-4a84-93c9-198b53596824


Epoch 200: Avg Loss = 0.4137

=== Tuning for 2024-03-28 ===


[I 2025-05-24 17:52:41,741] Trial 0 finished with value: -0.13522481421629587 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.13522481421629587.
[I 2025-05-24 17:52:42,125] Trial 1 finished with value: -0.3568883736928304 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3568883736928304.
[I 2025-05-24 17:52:42,498] Trial 2 finished with value: -0.11569659660259883 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3568883736928304.
[I 2025-05-24 17:52:42,867] Trial 3 finished with value: -0.12725848952929178 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3568883736928304.
[I 2025-05-

Training with params: {'reg_lambda': 0.00042013484629525313, 'div_lambda': 0.09890985260213801, 'lr': 0.00123528200251081}
Epoch 50: Avg Loss = 0.4228
Epoch 100: Avg Loss = 0.4110
Epoch 150: Avg Loss = 0.4112


[I 2025-05-24 17:53:07,207] A new study created in memory with name: no-name-adfcd740-a71b-441f-89ef-6a25ee178fa9


Epoch 200: Avg Loss = 0.4188

=== Tuning for 2024-06-28 ===


[I 2025-05-24 17:53:07,448] Trial 0 finished with value: -0.16711470981438956 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.16711470981438956.
[I 2025-05-24 17:53:07,820] Trial 1 finished with value: -0.3529793620109558 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.3529793620109558.
[I 2025-05-24 17:53:08,200] Trial 2 finished with value: -0.12550686299800873 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.3529793620109558.
[I 2025-05-24 17:53:08,591] Trial 3 finished with value: -0.13989011943340302 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.3529793620109558.
[I 2025-05-

Training with params: {'reg_lambda': 0.00037439493312826504, 'div_lambda': 0.099364353125051, 'lr': 0.0011897638792114668}
Epoch 50: Avg Loss = 0.4519
Epoch 100: Avg Loss = 0.4179
Epoch 150: Avg Loss = 0.4359


[I 2025-05-24 17:53:32,001] A new study created in memory with name: no-name-3cf4ddfd-5ce0-495e-bc33-5203d0d809e3


Epoch 200: Avg Loss = 0.4159

=== Tuning for 2024-09-30 ===


[I 2025-05-24 17:53:32,240] Trial 0 finished with value: -0.17549225687980652 and parameters: {'reg_lambda': 2.019450244810227e-05, 'div_lambda': 0.013172711156111799, 'lr': 0.00020664632886986634}. Best is trial 0 with value: -0.17549225687980652.
[I 2025-05-24 17:53:32,620] Trial 1 finished with value: -0.37370214859644574 and parameters: {'reg_lambda': 0.007831784183433456, 'div_lambda': 0.08165034948781692, 'lr': 0.00041256046816885543}. Best is trial 1 with value: -0.37370214859644574.
[I 2025-05-24 17:53:32,999] Trial 2 finished with value: -0.12894931435585022 and parameters: {'reg_lambda': 0.001010373285001627, 'div_lambda': 1.9418001640357832e-05, 'lr': 6.387294494138671e-05}. Best is trial 1 with value: -0.37370214859644574.
[I 2025-05-24 17:53:33,384] Trial 3 finished with value: -0.15986616412798563 and parameters: {'reg_lambda': 0.0009989183779852138, 'div_lambda': 0.005210986937158464, 'lr': 0.0025776094086318252}. Best is trial 1 with value: -0.37370214859644574.
[I 2025

Training with params: {'reg_lambda': 0.00020847533144764297, 'div_lambda': 0.09917491746255934, 'lr': 0.0011501797296916778}
Epoch 50: Avg Loss = 0.4238
Epoch 100: Avg Loss = 0.4410
Epoch 150: Avg Loss = 0.4315
Epoch 200: Avg Loss = 0.4254

=== Average Optimal Parameters ===
Regularization Lambda: 0.001412
Diversification Lambda: 0.099556
Learning Rate: 0.000788

Backtest completed!




Using random seed of 7 should converge with:


=== Average Optimal Parameters ===

Regularization Lambda: 0.001412

Diversification Lambda: 0.099556

Learning Rate: 0.000788

Backtest completed!

In [7]:
optuna.visualization.plot_optimization_history(study)

In [8]:

optuna.visualization.plot_param_importances(study)